src/cargo/util/process_builder.rs

```diff
@@ -107,6 +107,63 @@ impl ProcessBuilder {
         }
     }
 
+    pub fn exec_with_streaming<F1, F2>(&self, mut on_stdout_line: F1, mut on_stderr_line: F2)
+        -> Result<Output, ProcessError>
+        where F1: FnMut(&str), F2: FnMut(&str)
```

Comment:
For big functions like this I tend to try to keep them concrete (to avoid codegen bloat), perhaps these could both be `&mut FnMut(&str)` as arguments?